src/wrappers/themis/jsthemis/test/test.js

```diff
@@ -13,173 +27,409 @@ function expect_message(message) {
-	    	          expect_code(addon.INVALID_PARAMETER));
-	    server_session = new addon.SecureSession(valid_id, keypair.private(), function(){return null});
-	    assert.throws(function(){server_session.unwrap(empty_id)}, expect_code(addon.INVALID_PARAMETER));
-	})
-        it("callback behavior", function(){
-            message = new Buffer("This is test message")
-            server_id = new Buffer("server")
-            server_keypair = new addon.KeyPair()
-            client_id = new Buffer("client")
-            client_keypair = new addon.KeyPair()
-            broken_id = new Buffer("broken")
-            missing_id = new Buffer("missing")
-            unknown_id = new Buffer("unknown")
-
-            new_server_session = function() {
-                return new addon.SecureSession(server_id, server_keypair.private(), function(id){
-                    if(id.toString() == client_id.toString())
-                        return client_keypair.public()
-                    else if(id.toString() == missing_id.toString())
-                        return null
-                    else if(id.toString() == broken_id.toString())
-                        return 42
-                })
+    describe('Secure Session', function() {
+        let emptyArray = Buffer.from('')
+        let randomID = Buffer.from('random')
+        let clientID = Buffer.from('client')
+        let serverID = Buffer.from('server')
+        let messageA = Buffer.from('testing, testing')
+        let messageB = Buffer.from('everything seems to be in order')
+
+        function makeKeyMaterial() {
+            let keys = {}
+            keys.client = new addon.KeyPair()
+            keys.server = new addon.KeyPair()
+            keys.clientCallback = function(id) {
+                if (id.equals(serverID)) {
+                    return keys.server.public()
+                }
+                return null
+            }
+            keys.serverCallback = function(id) {
+                if (id.equals(clientID)) {
+                    return keys.client.public()
+                }
+                return null
             }
+            return keys
+        }
 
-            server_callback = function(id){
-                if(id.toString() == server_id.toString())
-                    return server_keypair.public()
+        it('establishes communication', function() {
+            let keys = makeKeyMaterial()
+            let client = new addon.SecureSession(clientID, keys.client.private(), keys.clientCallback)
+            let server = new addon.SecureSession(serverID, keys.server.private(), keys.serverCallback)
+
+            let data = client.connectRequest()
+            let peer = server
+            while (!(client.isEstablished() && server.isEstablished())) {
+                data = peer.unwrap(data)
+                if (peer == server) {
+                    peer = client
+                } else {
+                    peer = server
```

Comment:
this is a very smart and hacky way to establish async communication in sync piece of code 😂